src/Swift.Runtime/src/Library/ARC.cs

```diff
@@ -0,0 +1,167 @@
+// Copyright (c) Microsoft Corporation.
+// Licensed under the MIT License.
+
+using System;
+using System.Runtime.InteropServices;
+
+namespace Swift.Runtime;
+
+/// <summary>
+/// ARC is a class containing p/invokes for Swift reference counting and memory management.
+/// </summary>
+public static class ARC {
+    const string swiftCore = "libswiftCore.dylib";
+
+    /// <summary>
+    /// Retain a heap-allocated Swift object
+    /// </summary>
+    /// <param name="p">Pointer to an unmanaged Swift object, must be non-null.</param>
+    [DllImport (swiftCore)]
+    static extern void swift_retain (IntPtr p);
+
+    /// <summary>
+    /// Retain a heap-allocated Swift object
+    /// </summary>
+    /// <param name="p">Pointer to an unmanaged Swift object, must be non-null.</param>
+    /// <returns>The pointer passed in.</returns>
+    /// <exception cref="ArgumentOutOfRangeException">Throws if p is null</exception>
+    public static IntPtr Retain (IntPtr p)
+    {
+        if (p == IntPtr.Zero)
+            throw new ArgumentOutOfRangeException (nameof (p));
+        swift_retain (p);
+        return p;
+    }
+
+    /// <summary>
+    /// Check to see if a pointer is in the process of being deallocated.
+    /// </summary>
+    /// <param name="p">A non-null pointer to an unmanaged Swift object</param>
+    /// <returns>True if and only if the pointer in the process of being deallocated.</returns>
+    [DllImport (swiftCore)]
+    [return: MarshalAs (UnmanagedType.I1)]
+    static extern bool swift_isDeallocating (IntPtr p);
+
+    /// <summary>
+    /// Check to see if a pointer is in the process of being deallocated.
+    /// </summary>
+    /// <param name="p">A non-null pointer to an unmanaged Swift object</param>
+    /// <returns>True if and only if the pointer in the process of being deallocated.</returns>
+    /// <exception cref="ArgumentOutOfRangeException">Throws if p is null</exception>
+    public static bool IsDeallocating (IntPtr p)
```

Comment:
Is this call thread-safe?

Reply:
It looks like it. All the memory calls that reference the bits are protected with `std::atomic`